스트리머 및 SNS 선택 기능 구현

june0122 opened the pull request:
### 기능 추가
- [x] RoomDB에 Default StreamerList 정의
- [x] streamer, sns 리사이클러뷰에서 단일 아이템만 선택 가능하도록 구현
- [x] 선택한 스트리머의 데이터만 content 리사이클러뷰에 표시
- [x] 선택한 sns의 데이터만 content 리사이클러뷰에 표시
- [x] 유튜브, 트위치 영상의 Duration Time을 파싱하여 표시

### 리팩토링
- [x] Hilt 적용
- [x] DataBinding 적용
- [x] DiffUtil -> DiffUtil + ListAdapter를 이용한 방식으로 변경
- [x] Adapter 내부에서 처리하던 선택 상태를 ViewModel 내부에서 처리하도록 변경

june0122 committed:
chore: gradle 버전 업데이트

june0122 committed:
style: Streamer 아이템 클릭 시의 백그라운드 색상 추가

june0122 committed:
refactor: Collapsing Toolbar의 ScrollFlags 변경

'scroll|enterAlways'로 변경하여 스크롤을 맨위로 올리지 않아도
중간에 스크롤해도 Collapsing Toolbar를 볼 수 있도록 변경

june0122 committed:
style: streamer 아이템의 좌우 패딩값 변경

june0122 committed:
refactor: default streamerList를 DB에 저정하기 위해 Entity 추가 정의

june0122 committed:
feature: Room Database로부터 default data를 받아오기

june0122 committed:
feat: Streamer 리사이클러뷰 아이템 선택 효과 및 SmoothScroller 적용

june0122 committed:
chore: 매니페스트에 application name 정의 추가

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ContentsApplication.kt:
```diff
@@ -0,0 +1,13 @@
+import kotlinx.coroutines.SupervisorJob
+
+class ContentsApplication: Application() {
+    private val applicationScope = CoroutineScope(SupervisorJob())
```
질 1) GlobalScope과 CoroutinesScope의 차이가 뭘까요?
질 2) SupervisorJob()과 Job()이 있는데 어떤 차이가 있을까요?

june0122 replied:
답 1) GlobalScope는 어플리케이션과 수명 주기를 함께하는 싱글톤이지만 CoroutinesScope는 생성한 객체가 메모리에서 해제가 되면 함께 파괴가 됩니다.
답 2) SupervisorJob이 아닐 경우엔 예외가 발생했을때 자식 코루틴이 취소될 경우 부모 코루틴까지 취소가 전파가 되는데, SupervisorJob일 경우엔 취소가 부모 코루틴에 전파되지 않는 것으로 알고 있습니다. 아직 코루틴의 Exception과 SupervisorJob에 대해서는 자세히 공부하지 못해서 더 학습하여야 할 것 같습니다.

theo-f-lab replied:
답 1 - 질문 2) 그럼 여기서 작성하신 코드는 어떻게 동작할까요?

june0122 replied:
Repository와 Database의 인스턴스가 싱글톤으로 생성이 됩니다.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ContentsApplication.kt:
```diff
@@ -0,0 +1,13 @@
+
+class ContentsApplication: Application() {
+    private val applicationScope = CoroutineScope(SupervisorJob())
+    private val database by lazy { ContentRoomDatabase.getDatabase(this, applicationScope) }
```
Coroutines을 주입하고 있지만 초기화 데이터를 위해 사용하시는걸로 보이는데요. 초기화 할 수 있는 부분을 함수로 호출하고, GlobalScope 처럼 사용되는 CoroutineScope()을 주입하지 않는게 더 좋을것 같아요.

질문 ) Room의 onCreate는 언제 탈까요?

june0122 replied:
Database가 생성될 때 Room의 onCreate가 실행될 것 같습니다.

말씀하시는 Room의 onCreate가 아래의 onCreate를 말씀하시는 것일까요?

<img width="898" alt="image" src="https://user-images.githubusercontent.com/39554623/162653475-8ecc93a5-13ef-4202-99ec-0e65065581d2.png">

theo-f-lab replied:
네 룸의 콜백 onCreate 입니다.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/data/repository/ContentRepository.kt:
```diff
@@ -0,0 +1,9 @@
+import kotlinx.coroutines.flow.Flow
+
+class ContentRepository(contentDao: ContentDao) {
+    val isedolStreamers: Flow<List<StreamerEntity>> = contentDao.getStreamers()
```
변수명이 isedol 일 이유가 있을까요?

june0122 replied:
A팀 소속의 스트리머들만 보여주는 화면, B팀 소속의 스트리머들만 보여주는 화면, C팀 소속의 보여주는 화면 ... 식으로 구성할 예정이라 aStreamers, bStreamers, cStreamers와 같은 형태로 DB에 default data 형식으로 저장할 예정인데 A팀의 이름이 이세돌이기에 `isedolStreamers`라는 변수명을 사용하였습니다.

theo-f-lab replied:
변수명을 명확하게 하실 필요가 있는 부분입니다. 제가 isedol인 이유를 질문하였는데, 많은 설명이 함께 따라야 하고, 이를 다른 사람이 또 오면 또 설명을 해야겠죠.
적절해 보이지 않는 변수명에 해당될 수 있습니다. 조금 더 고민해보고 수정해보시면 좋을것 같습니다.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ContentsApplication.kt:
```diff
@@ -0,0 +1,13 @@
+class ContentsApplication: Application() {
+    private val applicationScope = CoroutineScope(SupervisorJob())
+    private val database by lazy { ContentRoomDatabase.getDatabase(this, applicationScope) }
+    val repository by lazy { ContentRepository(database.contentDao()) }
```
이 앱에서는 이 레파지토리 말고는 사용할 계획이 없나요?
하나 이상 추가되면 어떻게 작업 하실지 궁금합니다.

june0122 replied:
처음 구조를 대략적으로 그려봤을때는 하나의 레파지토리만 필요하다고 생각되어서 2개 이상의 레파지토리는 고려해보지 못했습니다..

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/HomeFragment.kt:
```diff
@@ -8,25 +8,45 @@ import androidx.databinding.DataBindingUtil
+    }
     private val contentListAdapter: ContentListAdapter = ContentListAdapter()
-    private val streamerListAdapter: StreamerListAdapter = StreamerListAdapter()
+    private val streamerListAdapter: StreamerListAdapter by lazy {
```
질문) lazy와 lateinit의 차이가 뭘까요?

june0122 replied:
`lazy`는 immutable 프로퍼티에서만 사용이 가능하고 초기화가 보장이 되는 반면, `lateinit`은 mutable 프로퍼티에서 사용이 가능하고 초기화가 보장이 되지 않고 원시 타입에서 사용이 불가능합니다.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/HomeFragment.kt:
```diff
@@ -49,6 +70,10 @@ class HomeFragment : Fragment() {
             contentListAdapter.updateUserListItems(it)
         }
+
+        homeViewModel.isedolStreamers.observe(requireActivity()) { streamers ->
```
여기서 생성한 ViewModel은 ActivityScope의 ViewModel인가요? Fragment의 Scope인가요?

ViewModel 초기화 코드상 fragment 스쿱의 ViewModel인데, activity를 사용해 observe 할 경우 어떻게 동작할까요?

june0122 replied:
Fragment Scope의 ViewModel이라 LifecycleOwner를 `viewLifecycleOwner`로 변경하여야 합니다!

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/adapter/StreamerListAdapter.kt:
```diff
@@ -2,42 +2,19 @@ package com.june0122.wakplus.ui.home.adapter
-        ItemStreamerBinding.inflate(LayoutInflater.from(parent.context), parent, false).also {
-            return StreamerListViewHolder(it)
+        ItemStreamerBinding.inflate(LayoutInflater.from(parent.context), parent, false).also { binding ->
+            return StreamerListViewHolder(binding, listener)
```
also의 의미가 무엇일까요?

june0122 replied:
수신 객체를 변경하지 않고 also 블록 내부에서 수신 객체를 그대로 사용하기 위해 사용합니다. 그렇기 때문에 swap 함수를 구현할 때도 also를 사용할 수 있습니다.

theo-f-lab replied:
그럼 이 코드에서는 also가 안맞지 않을까요? also { binding -> } binding이 원래 리턴

이 코드에서는 alos가 아닌 람다 표현 그 자체가 return 할 수 있는게 더 좋겠죠.
```
return ItemStreamerBinding.inflate(LayoutInflater.from(parent.context), parent, false).run {
           StreamerListViewHolder(binding, listener)
}
```

june0122 committed:
feat: 리사이클러뷰에서 하나의 항목만 선택 가능하도록 클릭 이벤트 추가

june0122 committed:
refactor: DiffCallback을 별도의 클래스로 분리

june0122 committed:
refactor: SNS별 ID를 `IdSet` 데이터 클래스를 통해 저장하도록 수정

june0122 committed:
feat: 선택한 스트리머의 데이터만 content 리사이클러뷰에 표시하는 기능 추가

june0122 committed:
feat: 영상의 duration을 보여주는 TextView 추가

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/HomeFragment.kt:
```diff
@@ -9,24 +9,53 @@ import androidx.fragment.app.Fragment
+        StreamerListAdapter(object : StreamerClickListener {
+            override fun onStreamerClick(position: Int) {
+                configureSmoothScroller(position)
+                streamerListAdapter.selectSinglePosition(position)
```
데이터를 직접 바꾸면 이 데이터를 관리하는 곳에선 어떻게 반영이 되는건가요?
최종 데이터가 adapter에서만 관리되는거라면 상관은 없겠지만

viewModel에서 하면 select > unselect를 처리하고, 이에 대한 이벤트 처리 역시 diffutil로 처리할 수 있고, select/unselect에 따른 함수 구분도 거기서 바로 할 수 있을것 같은데요.

요렇게 설계하신 이유가 뭘까요?
아침에 조언 드린 부분으론 viewModel에서 하시면 adapter 코드 수정없이도 가능한 부분일것으로 생각했는데 어떠한 부분에 대한 고민으로 이렇게 작성하시게 된것인지 궁금하네요.

june0122 replied:
@theo-f-lab 이전 북마크의 사례와 달리 데이터베이스에 저장될 필요가 없고, 뷰에서 position 값만을 받아와서 해당 item이 가지고 있는 id값을 함수에 넘겨주기만 하는 상황이라 데이터가 adapter에서만 관리되면 된다고 판단하였습니다.

`Streamer`에 `isSelected`와 같은 Boolean 프로퍼티를 추가로 정의하고 LiveData로 `isSelected`를 관리하는건 불필요하게 작업이 커지는 것 같아 위와 같이 처리하였습니다.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/HomeFragment.kt:
```diff
@@ -9,24 +9,53 @@ import androidx.fragment.app.Fragment
+                    homeViewModel.collectStreamerContents(streamerListAdapter[position].idSet)
+                }
+
+                homeViewModel.contents.observe(requireActivity()) {
```
이 코드는 어떻게 동작할까요? onStreamClick()에 있어보이는데요. 
제가 생각하기엔 streamClick()이 일어날때마다 이 함수가 새로 등록되고, updateuserListItems() 역시 N본 호출될걸로 생각이 드네요.

june0122 replied:
옵저빙 코드를 클릭 이벤트 내부에 중복해서 작성하는 실수를 했네요..!! 수정하겠습니다.

theo-f-lab replied:
자주 실수할 수 있는 부분은 나중에 글로 정리해두시면 좋을것 같아요.! 화이팅입니다.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/HomeFragment.kt:
```diff
@@ -57,13 +87,13 @@ class HomeFragment : Fragment() {
+            this.layoutManager = horizontalLayoutManager
             adapter = streamerListAdapter
-            addItemDecoration(StreamerItemDecoration(largePx))
+            addItemDecoration(StreamerItemDecoration(normalPx))
```
변수명이 large > normal로 변경되었는데, R.dmine. 만 변경되고, 나머지의 값은 변경 안되는게 좀 더 좋을것 같다는 생각이 드는데요. 그렇다면 명확한 변수명을 가지는게 좋겠죠.

june0122 replied:
```xml
<dimen name="margin_extra_large">32dp</dimen>
<dimen name="margin_large">24dp</dimen>
<dimen name="margin_normal">16dp</dimen>
<dimen name="margin_small">8dp</dimen>
<dimen name="margin_extra_small">4dp</dimen>
```

마진을 줄일 필요가 있어서 기존에 dimens에 정의되어있던 margin_large에서 margin_normal로 변경한 코드인데 어떠한 부분이 문제가 되는 것인지 잘 이해하지 못했습니다..

theo-f-lab replied:
네이밍 규칙을 말한것이였어요. 이름이 명확했다면 굳이 안바꾸고, dimen. 호출 부분만 변경될 수 있으니깐요.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/HomeViewModel.kt:
```diff
@@ -53,49 +53,68 @@ class HomeViewModel : ViewModel() {
+            contents
+        }
+
+        return contentData.await()
```
launch에서 async/await으로 변경되었는데요. 어떠한 차이가 있을까요?

june0122 replied:
launch는 `Job`을 반환하지만 async는 `Deferred<T>`를 반환하기에 `await()`를 통해 async 코루틴 빌더 내부의 결과값을 사용할 수 있습니다.

theo-f-lab replied:
launch와 async/await에서의 실행은 어느 시점에 동작할까요?
호출 직전? 호출 하고 사용시점?(await 시점)

june0122 replied:
launch와 async 둘 다 호출 직전에 동작하고, async의 생성자에 `start = CoroutineStart.LAZY`를 인수로 넘겨줄 경우에는 사용 시점에 동작하는 것으로 이해하고 있습니다.

theo-f-lab replied:
💯

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/adapter/StreamerListAdapter.kt:
```diff
@@ -46,5 +30,41 @@ class StreamerListAdapter : RecyclerView.Adapter<StreamerListViewHolder>() {
+        notifyItemChanged(selectedPosition)
+        previousSelectedPosition = selectedPosition
+        selectedPosition = adapterPosition
+        notifyItemChanged(selectedPosition)
```
이 코드는 diffutil을 힘들게 쓰고, 이를 의미 없이 만드는 코드로 보여지네요. diffUtil을 이해했다면 굳이 이 코드가 필요치 않겠죠.

notifyItemChanged의 동작 방법을 알려주세요.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/utils/CenterSmoothScroller.kt:
```diff
@@ -0,0 +1,10 @@
+import android.content.Context
+import androidx.recyclerview.widget.LinearSmoothScroller
+
+open class CenterSmoothScroller(context: Context?) : LinearSmoothScroller(context) {
```
open의 의미가 뭘까요?

june0122 replied:
코틀린의 클래스는 기본적으로 `final`인데 `open` 키워드를 통해 클래스의 상속 및 오버라이드가 가능해집니다.

theo-f-lab replied:
이 클래스의 경우 확장이 필요한걸까요? 굳이 필요치 않으면 open을 붙여두는건 큰 의미는 없습니다.

june0122 replied:
함수 재정의가 필요할 줄 알고 open으로 작성했었는데 현재 시점에선 불필요한게 맞네요..!

theo-f-lab replied:
사실 상관은 없지만, kotlin의 장점이 닫아두고, 필요할대 열어두는 형태이다보니 확장을 잘 정의 하셔야 합니다!

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/utils/DiffCallback.kt:
```diff
@@ -0,0 +1,25 @@
+
+import androidx.recyclerview.widget.DiffUtil
+
+open class DiffCallback<T>(
```
이 클래스는 어떠한 용도인가요?

june0122 replied:
기존엔 DiffCallback 클래스가 어댑터 클래스(`ContentListAdapter`, `StreamerListAdapter`)가 정의된 파일 내에 함께 정의되어 있었는데 이것을 분리하고 내부 함수를 override해서 사용할 수 있도록 하였습니다.

theo-f-lab replied:
근데 작성하신 코드를 보면 override의 의미가 없어서요. super를 부르는것도 아니니... 그냥 이 코드 의미가 없어보입니다.

june0122 replied:
2개의 어댑터 클래스 내부에서 `areItemsTheSame(...)`을 재정의하기 위해 위와 같이 작성하였는데 다른 방법이 있을까요?

#### StreamerListAdapter

```kotlin
  fun updateStreamerListItems(items: List<StreamerEntity>) {
        val diffCallback = object : DiffCallback<StreamerEntity>(streamerList, items) {
            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = oldItems[oldItemPosition]
                val newItem = newItems[newItemPosition]
                return oldItem.name == newItem.name
            }
        }
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        streamerList.clear()
        streamerList.addAll(items)
        diffResult.dispatchUpdatesTo(this)
    }
```

#### ContentListAdapter

```kotlin
    fun updateUserListItems(items: List<ContentData>) {
        val diffCallback = object : DiffCallback<ContentData>(contentList, items) {
            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                val oldItem = oldItems[oldItemPosition]
                val newItem = newItems[newItemPosition]

                return if (oldItem is TwitchVideoEntity && newItem is TwitchVideoEntity)
                    oldItem.twitchVideoInfo.id == newItem.twitchVideoInfo.id
                else if (oldItem is YoutubeVideoEntity && newItem is YoutubeVideoEntity)
                    oldItem.youtubeVideoInfo.id == newItem.youtubeVideoInfo.id
                else oldItem == newItem
            }
        }
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        contentList.clear()
        contentList.addAll(items)
        diffResult.dispatchUpdatesTo(this)
    }
```

theo-f-lab replied:
DiffUtil 부분만을 확장하는게 아닌 아래 contentList.clear(), contentList.addAll, diffResult.dispatchUpdates 등도 모두 자동으로 처리하는편이 더 좋을거에요.
사실 공통 코드를 줄이고 싶다고 하시면 이 부분까지를 모두 함수로 빼는 방법

더 좋은 방법은  ListAdapter를 추가로 학습하셔서 사용하시는게 더 좋은 방법입니다.
https://youngest-programming.tistory.com/474

june0122 committed:
chore: Twitch User ID 변경 적용

Twitch User ID가 변경되는 사례가 있으므로 실시간으로 ID를 불러오도록
추후 수정 필요

june0122 committed:
refactor: 유튜브 API 할당량 최적화 작업

유튜브 API로부터 채널의 동영상 정보를 조회하는 작업을 채널의 재생목록을 모두 조회하여
동영상 정보를 가져오는 방식으로 변경하여 할당량을 기존의 1/6 ~ 1/9 수준으로 최적화

june0122 committed:
refactor: CenterSmoothScroller 클래스를 재정의하지 않고 사용하도록 변경

june0122 committed:
refactor: scope 함수를 사용하지 않는 코드로 변경

june0122 committed:
refactor: Streamer 선택 기능을 어댑터 내부에서가 아닌 뷰모델에서 처리

어댑터 내부에서 구현한 스트리머 리사이클러뷰의 단일 선택 기능을 LiveData를
이용하고, 뷰모델에서 처리하도록 하여 좀 더 MVVM 패턴에 알맞은 형태로 수정

june0122 committed:
resource: 유튜브 프로필 이미지 리소스 추가

june0122 committed:
refactor: strings.xml에 정의된 값을 사용하여 text를 표시하도록 변경

june0122 committed:
refactor: 주 생성자의 인수로 `getContext()` 대신 `requireContext()` 전달

코틀린에서 Null이 아닌 Context를 전달할 때 `requireContext()`를 사용할 수 있다.
`requireContext()`를 통해 Context가 Null이 아님을 보장할 수 있다.

june0122 committed:
refactor: 주생성자에 binding 프로퍼티를 정의 및 초기화하여 불필요한 맵핑 제거

june0122 committed:
refactor: init 블록의 위치 변경 및 Flow 중간 연산자를 적절하게 변경

streamers 관련 코드의 위치를 바로 찾을 수 있도록 인접한 곳으로 init 블록 위치 변경
Flow를 변환하는 중간 연산자(intermediate operator)로 map을 사용하였는데
순회하며 값을 변환하는 것이 아니라 순회하며 리스트에 추가하는 작업이므로 onEach 중간 연산자를
사용하는 것이 더 적절하다.

_streamers.value 값을 변경할 때 init 블록 내부에선 null인 데이터가 올 수 있기 때문에
엘비스 연산자 처리를 해주었지만, onStreamerClick 내부에선 null인 데이터를 고려하지 않아도
되기에 엘비스 연산자 처리를 생략.

조건문 처리를 간결하게 수정.

june0122 committed:
refactor: `getYoutubeVideos(...)` 함수의 가독성을 고려하여 코드 수정

june0122 committed:
fix: privacyStatus가 public이 아닌 private(비공개) 처리된 영상 예외 처리

june0122 committed:
refactor: DiffUtil만 사용하던 코드에서 ListAdapter 추가로 적용

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/adapter/StreamerListAdapter.kt:
```diff
@@ -2,49 +2,25 @@ package com.june0122.wakplus.ui.home.adapter
-        ),
-    )
+class StreamerListAdapter(private val listener: StreamerClickListener) :
+    ListAdapter<StreamerEntity, StreamerListViewHolder>(StreamerDiffCallback()) {
```
ListAdapter를 적용하셨는데, RecyclerView.Adapter와 어떠한 차이가 있을까요?

june0122 replied:
*ListAdapter*를 사용할 경우
- adapter 내부에 별도의 list를 정의하지 않고도 `currentList`를 이용하여 리스트의 아이템에 접근할 수 있습니다.
- `getItemCount()` 오버라이드 메서드를 따로 정의하지 않아도 됩니다.
-  *ListAdapter*를 사용하지 않고 DiffUtil을 구현할 때는 `calculateDiff()`로 변경된 아이템을 감지하여 `dispatchUpdatesTo(adapter)`를 통해 지정한 어댑터에 업데이트 이벤트를 전달하는 코드를 직접 작성해야 했는데, ListAdapter는 내부적으로 이러한 구현들을 처리해주고 비교 연산을 백그라운드 스레드에서 처리하도록 도와주는 *AsyncListDiffer*의 Wrapper이기 때문에 코드가 훨씬 간결해집니다. 
- `notify~` 시리즈의 메서드를 사용하지 않고도 `submitList(...)`를 통해 변경사항들을 업데이트할 수 있습니다.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/home/adapter/StreamerListAdapter.kt:
```diff
@@ -2,49 +2,25 @@ package com.june0122.wakplus.ui.home.adapter
-        ),
-    )
+class StreamerListAdapter(private val listener: StreamerClickListener) :
+    ListAdapter<StreamerEntity, StreamerListViewHolder>(StreamerDiffCallback()) {
```
ListAdapter를 적용하셨는데, RecyclerView.Adapter와 어떠한 차이가 있을까요?

june0122 committed:
fix: 조회수가 소수점 첫째 자리 밑으로도 표시되는 문제 수정

june0122 committed:
refactor: ViewBinding에서 DataBinding으로 마이그레이션

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/BindingAdapters.kt:
```diff
@@ -0,0 +1,68 @@
+    @BindingAdapter("twitchThumbnailUrl")
+    fun setTwitchThumbnailUrl(view: ImageView, imageUrl: String) {
+        imageUrl.replace("%{width}x%{height}", RES_STANDARD).let { convertedUrl ->
+            view.load(convertedUrl) {
```
이 부분의 코드가 동일하니 조합만하고, 다른.바인딩 함수로 전달하는것도 방법이에요.
아니면 데이터 던질때 가공하고 이 함수는 하나로 관리하는게 오히려 관리 관점에 명확한 구분을 할.수도 있고요.

theo-f-lab commented on app/src/main/java/com/june0122/wakplus/ui/BindingAdapters.kt:
```diff
@@ -0,0 +1,68 @@
+    @BindingAdapter("isSelected")
+    fun setStreamerLayoutBackground(view: ConstraintLayout, isSelected: Boolean) {
+        if (isSelected) view.setBackgroundResource(R.color.Primary50)
+        else view.setBackgroundResource(R.color.transparent)
```
컬러만.다르니 컬러만 표현식으로 정의하고
그.변수값알 set하는게.더 좋겠죠.

june0122 replied:
```xml
android:background="@{streamer.selected ? @color/Primary50 : @color/transparent}"
```
눈에도 바로 들어오고 따로 바인딩 함수를 정의 안해도 되서 너무 좋네요!

theo-f-lab commented on app/src/main/res/layout/item_streamer.xml:
```diff
@@ -5,14 +5,27 @@
         android:layout_height="wrap_content"
         android:clickable="true"
         android:focusable="true"
+        android:onClick="@{()->streamerClickListener.onStreamerClick(position)}"
```
이런 경우 포지션을.던질스도 있지만 아이템을.던지는 것도 방법입니다.

june0122 committed:
refactor: 바인딩 함수의 공통된 부분 분리 및 표현식으로 변경

바인딩 함수를 정의하지 않고 바인딩 표현식의 삼항 연산자로 구현 가능한 부분 수정

june0122 committed:
refactor: ItemDecoration에 전달할 px값의 변수명을 명확하게 변경

june0122 committed:
resource: 라이브 방송 중 썸네일이 표시되지 않는 트위치용 Placeholder 이미지 추가

june0122 committed:
style: 동영상 제목의 maxLine 설정 및 줄임말(ellipsize) 옵션 추가

june0122 committed:
refactor: `Context.getString()`을 통해 문자열 표시하도록 수정

june0122 committed:
style: 영상 업로드 기간 표시를 위한 문자열 추가 정의

june0122 committed:
feat: 트위치 썸네일 url이 존재하지 않을 때 표시할 placeholder 이미지 추가

june0122 committed:
refactor: ListAdapter 적용으로 인한 별도의 DiffCallback 클래스 제거

june0122 committed:
style: 파일명 변경으로 인한 삭제

june0122 committed:
feat: content들을 플랫폼 관계없이 최근 업로드 기준으로 정렬하는 메서드 추가

june0122 committed:
feat: 유튜브와 트위치 API로부터 받아 온 duration time을 변환하여 표시

june0122 committed:
feat: BottomNavigationView의 아이콘 및 텍스트 색상 지정

june0122 committed:
style: 문자열 리소스에 SNS의 이름들 추가

june0122 committed:
feat: SNS 선택 기능 구현 및 DB에 기본 SNS 리스트 저장

june0122 committed:
refactor: content 관련 메서드 분리 및 메서드명 변경

june0122 committed:
feat: contents 로딩 관련 코루틴 작업 취소 추가

june0122 committed:
refactor: 불필요한 `join()` 함수와 코루틴 생성 코드 제거

june0122 committed:
feat: Hilt 라이브러리를 통해 의존성 주입 적용
